File: storage/db-core/src/main/java/com/coffee/gu/coupon/CouponRepositoryImpl.java
package com.coffee.gu.coupon;

import com.coffee.gu.enums.EntityStatus;
import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

import static com.coffee.gu.coupon.QCouponEntity.couponEntity;

@Repository
public class CouponRepositoryImpl implements CouponRepository{

    private final JPAQueryFactory jpaQueryFactory;

    public CouponRepositoryImpl(JPAQueryFactory jpaQueryFactory) {
        this.jpaQueryFactory = jpaQueryFactory;
    }

    @Override
    public List<Coupon> findAllByIdIn(List<Long> couponIds) {
        return jpaQueryFactory.selectFrom(couponEntity)
                .where(
                        couponEntity.id.in(couponIds),
                        couponEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetch()
                .stream()
                .map(CouponEntity::toModel)
                .toList();
    }

    @Override
    public Optional<Coupon> findById(Long couponId) {
        return Optional.ofNullable(
                jpaQueryFactory.selectFrom(couponEntity)
                        .where(
                                couponEntity.id.eq(couponId),
                                couponEntity.entityStatus.eq(EntityStatus.ACTIVE)
                        )
                        .fetchFirst()
        ).map(CouponEntity::toModel);
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/coupon/CouponRepositoryImpl.kt
package com.coffee.gu.coupon

import com.coffee.gu.coupon.QCouponEntity.couponEntity
import com.coffee.gu.enums.EntityStatus
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import java.util.Optional

@Repository
class CouponRepositoryImpl(
    private val queryFactory: JPAQueryFactory,
) : CouponRepository {

    override fun findAllByIdIn(couponIds: List<Long>): List<Coupon> {
        return queryFactory.selectFrom(couponEntity)
            .where(
                couponEntity.id.`in`(couponIds),
                couponEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetch()
            .map { it.toModel() }
    }

    override fun findById(couponId: Long): Optional<Coupon> {
        return Optional.ofNullable(
            queryFactory.selectFrom(couponEntity)
                .where(
                    couponEntity.id.eq(couponId),
                    couponEntity.entityStatus.eq(EntityStatus.ACTIVE),
                )
                .fetchFirst(),
        ).map { it.toModel() }
    }
}
